### PR TITLE
Enable distance splits

### PR DESCRIPTION
**Pull request by neil-marcellini**
### Details

Enable splitting distance expenses. The goal is to make sure that the feature works for the mainline cases, that it doesn't break existing split flows, and that it's still hidden by the beta. There will be a number of small bugs that already exist on main and probably a few with this new feature. In order to keep things moving, these bugs will be listed and issues will be created to fix them with help from the community before removing the beta.

### Fixed Issues

$ https://github.com/Expensify/App/issues/36967
PROPOSAL: N/A

### Tests

**Known bugs on main:**
While testing please take these into consideration and ignore them since they were not caused by this PR.
- [ ] Chats will show blank and won't load report actions. Sometimes switching chats will fix this.
- [x] https://github.com/Expensify/App/issues/44692
  - [x] Now, receipt images don't load at all
- [ ] Splitting with an existing group chat shortly after sign in fails due to incorrect optimistic 1:1 chat ID, since the existing DM isn't in the allReports cache.
- [ ] global create allows submit expense P2P distance outside of beta

**New bugs to fix later**
Maybe new, maybe happen on main
- [ ] Split with new workspace, error, selected rate has been deleted
- [ ] Split with existing workspace, multiple previews displayed
- [ ] receipt image doesn't display in split details (related to other receipt issues on dev)

**A. Split distance with new accounts**
1. Sign into an account on the p2pDistanceRequests beta
2. Click the green plus, Split expense
3. Choose distance, enter 2 waypoints, click Next
4. Enter 2 new accounts and select them for the split
5. Verify that the split on the confirmation page looks good and makes sense, then save it
6. Verify that a new group chat is created and you are navigated to it
7. Verify the new split distance expense is displayed with a route receipt image, the type "Distance", an amount, merchant  in the correct distance format, the split participant avatars, and the "Your split" text.
8. Click on the split expense
9. Verify the split details page opens showing who its paid by, the participants, the route receipt, amount and description.
10. Click show more
11. Verify the merchant and date show
12. Go to a DM with one of the split participants
13. Verify the split expense shows with the route, "<user-email> owes:" text, amount, and merchant. Verify the amount is an even split of the total. Verify the distance is also a split of the total and multiplies times the rate to equal the split amount.
14. Click on it to open the transaction thread
15. Verify the details there look correct

**B. Workspace distance split**
1. Sign into an account on the p2pDistanceRequests beta
2. Split a distance expense with a workspace
3. Verify that the split on the confirmation page looks good and makes sense, then save it
4. Verify the split expense is created in the workspace chat
5. Verify the new split distance expense is displayed with a route receipt image, the type "Distance", an amount, merchant  in the correct distance format, the split participant avatars, and the "Your split" text.
7. Click on the split expense
8. Verify the split details page opens showing who its paid by, the participants, the route receipt, amount and description.
9. Click show more
10. Verify the merchant and date show

**C. Manual split with existing accounts**
1. Split an expense manually
3. Verify everything looks right

**D. Verify splits without the p2pDistanceRequests beta**
1. Sign into an account **not** on the p2pDistanceRequests beta, or disable the beta in the frontend code
2. Click green plus, split expense
3. Verify that only manual and scan appear as options

**E. Split with an existing group**
1. Create a group chat
2. Split a distance expense with them
3. Verify it's created properly

**F. Manual split new accounts**
1. Split an expense manually
2. Enter emails that haven't signed up yet
3. Verify everything looks right

- [ ] Verify that no errors appear in the JS console

### Offline tests

Skipping these for now since the existing tests are a lot to run through already. I'm pretty sure it will work well offline, but we can fine tune that while the feature is still under beta and before adding the [regression tests](https://docs.google.com/document/d/15GQhipcqOJ69MbFA54KRUcWxjICURHYcfTlAb3eaVEQ/edit#heading=h.j93ukuhgh4on) for the feature.

### QA Steps

Same as tests. If you find any issues please tag me in Slack and make sure they are not labeled as a deploy blocker, unless it's an issue that happens outside of the beta.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
I only tested on web because the changes are platform independent, and C+ will help test all platforms.
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

**A**

https://github.com/Expensify/App/assets/26260477/19cc0a0e-4de1-4e5d-831e-6d995c2d8eb6

**B**

https://github.com/Expensify/App/assets/26260477/65f233f0-bbc1-443d-93e4-aa1ed1fe0f1e

**C**

https://github.com/Expensify/App/assets/26260477/c8af9767-90a2-4379-9ab9-e3e2367c739d

**D**

https://github.com/Expensify/App/assets/26260477/763e19fa-46ed-48d6-9534-32aa6e34626e

**E**

https://github.com/Expensify/App/assets/26260477/57bede06-2084-418d-854e-fff1f8781210

**F**

https://github.com/Expensify/App/assets/26260477/f10a3843-89f8-4206-94ad-a7392baa3b9c

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by neil-marcellini** [comment]
@mananjadhav we will start reviewing this internally now, and I will let you know once the backend PRs are on staging.

**Comment by mananjadhav** [comment]
Okay sure. Let me know once this is ready for review. I'll get to it.

**Review by hayata-suenaga — DISMISSED** [review]
🟢 to me. I'd wait for @chiragsalian's approval

thank you for the work 😄

**Review by chiragsalian — DISMISSED** [review]
i havent tested but the code mostly LGTM

**Comment by neil-marcellini** [comment]
@mananjadhav this is ready for your review and testing now.

**Comment by mananjadhav** [comment]
Did a partial review, will finish the rest today.

**Review by mananjadhav — COMMENTED** [review]
Code looks good.

**Review by hayata-suenaga — APPROVED** [review]
🟢

File: src/libs/actions/IOU.ts
Comment (neil-marcellini) [reply src/libs/actions/IOU.ts:1990]: Yes, building the optimistic transaction will set the receipt and filename on the transaction. Previously the filename was only set from the receipt name when it has a source, defaulting to an empty string, and that's what it does now.

**Comment by neil-marcellini** [comment]
@mananjadhav would you please test and post the checklist? Please watch out for the existing bugs that I noted in the description, and note any new bugs you find and whether they are new or already present on main. Thanks!

**Comment by mananjadhav** [comment]
I did test this on Web, I'll test the rest and finish the checklist today.

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by mananjadhav** [comment]
I am getting weird issues for my recent tests. I disabled the strict mode.

**Can't edit amount**

https://github.com/user-attachments/assets/ae4cd0cb-d8cc-47a6-84e6-e7caad26d78d

**even in the split screen**

https://github.com/user-attachments/assets/9af15e8c-3c7a-49e2-8908-4e9a350b91ab

Endless loader for the Distance Split request
Screenshot 2024-07-30 at 1 08 56 AM

Composer got hidden after performing the split request and closing the error.

https://github.com/user-attachments/assets/3e53bd83-3248-4593-86dc-06c67e8fb443

**Comment by neil-marcellini** [comment]
Hi @mananjadhav. Sorry for the delay in getting back to this. I just made one small change for clean up, merged main, and now I think it's all working well. Please test again and let me know if you have any problems. Also please include reproduction steps for any bugs you find.


https://github.com/user-attachments/assets/b0d559c5-4f61-4320-bd84-11144c377011


https://github.com/user-attachments/assets/22b2aeb4-959a-4313-8e5e-905ef5adb583

While I wait for your next review I will re-test the bugs I found in this branch and on main and report them.

**Comment by mananjadhav** [comment]
thanks @neil-marcellini. I'll review and test this again.  I'll put in the reproduction steps for future reports but several of these looked unrelated to the PR.

**Comment by mananjadhav** [comment]
Starting on this now.

**Comment by mananjadhav** [comment]
@neil-marcellini We can skip for this one, but offline tests won't work. If we create an offline distance split, we don't have the amount and the splits throw an error: `The sum of splits must equal the total amount.`.
 
image

https://github.com/user-attachments/assets/5d1cf7b4-e521-4f36-8f03-fe0737009039

**Comment by mananjadhav** [comment]
Just finished testing different scenarios on web and everything else looks good. Highlighted the offline case in the previous comment.

**Comment by mananjadhav** [comment]
### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

**Distance Split**

https://github.com/user-attachments/assets/81fea8bc-865e-4a1b-b057-d9b3486b45cb

**Distance Split Other**

https://github.com/user-attachments/assets/afa83b96-ef0c-4c6e-bdea-f0f8877da7f4


**Offline Split -- doesn't work**
https://github.com/user-attachments/assets/0ade6a41-a5d3-4cc1-b4c6-66f97a300f15


</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/9050dcdf-cbe6-4c37-b46a-ea0ef520fe0b

</details>

**Review by mananjadhav — APPROVED** [review]

**Review by arosiclair — APPROVED** [review]
👍

**Comment by youssef-lr** [comment]
Looks like there is a bug where the QAB distance split uses 'SplitBill' endpoint when it should use 'CreateDistanceRequest' https://github.com/Expensify/App/issues/47021.

**Comment by pecanoro** [comment]
This PR also broke TrackExpense on DMs since it started calling the wrong endpoint: https://github.com/Expensify/App/issues/47042#issuecomment-2275466880

File: src/pages/iou/request/step/IOURequestStepConfirmation.tsx
Comment (jjcoffee) [inline src/pages/iou/request/step/IOURequestStepConfirmation.tsx:363]: Moving this `createDistanceRequest` call here caused an error in https://github.com/Expensify/App/issues/47042 when tracking a distance expense as we end up calling `CreateDistanceRequest` here instead of `trackExpense`.

**Pull request merged**
